### ISSUE TITLE
Alert for high server `network`/`ram`/`disk`/`cpu` usage "heartbeat" monitoring

### ISSUE THREAD

**Issue by 8bitUpsurge** [louislam/uptime-kuma#819]
### 🏷️ Feature Request Type

New Monitor

### 🔖 Feature description

Please add Heartbeat Monitoring just like in HetrixTools.

### ✔️ Solution

Add a new monitor type which shows server's network usage, ram usage, disk and cpu usage and gives alert when usage is close to 90% so people know it's time to upgrade or add a new node.

### ❓ Alternatives

HetrixTools

### 👀 Have you spent some time to check if this feature request has been raised before?

- [X] I checked and didn't find similar feature request

**Comment by bitUpsurge** [comment]
Nope, HetrixTool has Hearbeat Monitoring Under Uptime Monitoring Product.

**Comment by markdesilva** [comment]
Sounds like maybe you should be using HetrixTools then as you’re obviously a fan.

**Comment by rihards-simanovics** [comment]
> Nope, HetrixTool has Hearbeat Monitoring Under Uptime Monitoring Product.

What you are asking here is out of scope (perhaps for now), full stop.

Unless you are willing to code it yourself, wait until the UK developer does it.

Surely you understand that everyone who develops and contributes to a free and open-source product dedicates their free time to do so. The feature that you are asking for is from a paid product, there is a reason why it's paid, the money goes to a developer for their hard work.

**Comment by louislam** [comment]
Everyone relax🐻. 

Just follow one rule. If you love the suggestion, give a 👍.

Ignore it if you don't like it.

**Comment by zimbres** [comment]
I use https://www.netdata.cloud/

**Comment by rihards-simanovics** [comment]
@zimbres, this is amazing and Open Source. Hmm, I think I might have this running for my client reports... I will keep using the UK for internal services as those don't require reports generated. Thanks for sharing the tool name!

PS: Perhaps the UK might use some of the source code or take inspiration from that tool as it looks quite nice.

**Comment by ririko5834** [comment]
This should works like hetrixtools works, that you  can get also stats about ram, CPU, disk, network, etc. displayed in graph on status page.

**Comment by markdesilva** [comment]
Here we go again with Hetrix tools. 

I wonder if all these folks suggesting UK work like Hetrixtools just want the HT functions cos they want the unlimited HT functionality without paying for it.

Sounds like it doesn’t it? 🤷🏻‍♂️

**Comment by rihards-simanovics** [comment]
@ririko5834 the basic answer to your request is "maybe in the future".

UptimeKuma is a relatively simple uptime monitoring application running on NodeJS. not saying that NodeJS is a bad language. Still, I am nearly pointing out that a different language is more favourable due to performance requirements for what you are asking.

As @markdesilva pointed out, and I agree with them, if you favour Hetrix Tools, you need to support the developer by getting a paid plan. UptimeKuma may be an open-source project for now, but I'm sure that when the time comes, the author will also want to have their own paid plans alongside Open Source for those people who don't want to have a hustle of setting one up themselves.

That being said, keep in mind that, as pointed out by @ImmaZoni, to know the server's hardware status, the author will require developing and requesting to installing of a separate "companion" app on the server, which will push the CPU, RAM, etc. information to UptimeKuma. If the last stable release (v1.14.0) is anything to go by, the author wants this application to just run without any additional hoops to jump through (ref. to Cloudflare proxy functionality).

**Comment by InSelfControll** [comment]
> @ririko5834 the basic answer to your request is "maybe in the future".
> 
> UptimeKuma is a relatively simple uptime monitoring application running on NodeJS. not saying that NodeJS is a bad language. Still, I am nearly pointing out that a different language is more favourable due to performance requirements for what you are asking.
> 
> As @markdesilva pointed out, and I agree with them, if you favour Hetrix Tools, you need to support the developer by getting a paid plan. UptimeKuma may be an open-source project for now, but I'm sure that when the time comes, the author will also want to have their own paid plans alongside open-source for those people who don't want to have a hustle of setting one up themselves.
> 
> That being said, keep in mind that, as pointed out by @ImmaZoni, to know the server's hardware status, the author will require developing and requesting to installing of a separate "companion" app on the server, which will push the CPU, RAM, etc. information to UptimeKuma. If the last stable release (v1.14.0) is anything to go by, the author wants this application to just run without any additional hoops to jump through (ref. to Cloudflare proxy functionality).
> 
> EDIT:
> Almost forgot, @zimbres also noted that there is another open-source tool called **netdata** that you can use to monitor server hardware status.

Hey you don't have to install anything on the server just need to let uptime-kuma the option to connect via ssh to every server and get these info then parse it to the status page.

I have made a bash script that send details like this to my email once it pass the 75% disk usage same goes to the ram and CPU.

I just need to find the right way to send the data to uptime-kuma now for it to send it to me via telegram.

**Comment by rihards-simanovics** [comment]
@InSelfControll
> Hey you don't have to install anything on the server just need to let uptime-kuma the option to connect via ssh to every server and get these info then parse it to the status page.

Do you even realise how dangerous this is? Openly allow an application (of all things) to have access to a server via SSH? It's almost as if security holed don't exist. So now the hacker instead of hacking six of my servers only need to hack one of my servers and get SSH access to all the other servers.

The best and most secure way is to have a dedicated client application that would receive a request (be it via the web URL or else), process it and send a response JSON to an API on the UK side, or alternately just send the JSON data with an interval, so there is only one way communication from server to UK.

What you've proposed breaks the security best practises on so many levels.

UK - UptimeKuma

**Comment by InSelfControll** [comment]
This user doesn't need any permissions except df -h, free -m commands you always can minimized the commands of a user to only 1/2 commands or give the user limited ssh access to only send this commands via ssh @rihards-simanovics, you don't have to give fully login access to ssh so no security issues.
Today I have it automatically send to my email / telegram from each server.

**Comment by rihards-simanovics** [comment]
> This user doesn't need any permissions except df -h, free -m commands you always can minimized the commands of a user to only 1/2 commands or give the user limited ssh access to only send this commands

I did think of that, that being said it is still a very junky solution. Besides, it's already been mentioned in this discussion that much better paid applications are available. If you have enough servers to warrant am advanced system like that perhaps it's time to get the wallet out.

> This user doesn't need any permissions except df -h, free -m commands you always can minimized the commands of a user to only 1/2 commands or give the user limited ssh access to only send this commands via ssh @rihards-simanovics, you don't have to give fully login access to ssh so no security issues.
Today I have it automatically send to my email / telegram from each server.

Again, it's almost as if security holes don't exist, you are playing an extremely dangerous game by even allowing the potential hacker to login. Look I'm no security expert but I can guarantee you gaining access as a "limited user" is a first step to a full blown hack, so lets not.

**Comment by InSelfControll** [comment]
The other option is to fix the push passive monitor, and let users to send
custom messages in it.

Now the only message it sends is "ok" kinda useless message.

I want to send the script output via curl into the push passive monitor
instead just receiving "ok" message.

Now each of my VMS runs the script all the time and if the disk usage is
more then 85% I receive an email with the status and details about the disk
usage.

**Comment by markdesilva** [comment]
Maybe I don't quite understand your description, but the status msg can take any message. It does not accept the messages in quotes like "Service is up" but it will take URL spaces as in %20. Eg:

```
attl=`/usr/bin/ping -c 1 <UK server IP> | tail -1 | /usr/bin/cut -d"/" -f 5`

/usr/bin/curl -k "https://<UK server IP>:3001/api/push/XXXXXXXXX?msg=Service%20is%20up,%20ping%20time%20is%20$attl&ping=$attl"
```
As you can see, you can even pass variables (in this case the ping time to the UK server).

Then your UK will show this:

![uk-push-service-msg](https://user-images.githubusercontent.com/59959024/209102943-8dc8fba2-d1aa-41dd-8608-031abef792bc.jpg)

If you're using a linux machine, normal users (non root, no sudo) have access to to `cat /proc/cpu` info and `cat /proc/meminfo`, as well as `df` and can use `cut`, `sed`, `awk`, `grep` whatever info they need and pass it into the status message. For windows I think there are equivalent Powershell commands normal users can use to get the values for cpu usage, memory usage and disk usage (Get-Volume). 

Hope it works for you.

Cheers!


> The other option is to fix the push passive monitor, and let users to send custom messages in it. Now the only message it sends is "ok" kinda useless message. I want to send the script output via curl into the push passive monitor instead just receiving "ok" message. Now each of my VMS runs the script all the time and if the disk usage is more then 85% I receive an email with the status and details about the disk usage.

**Comment by InSelfControll** [comment]
> @markdesilva this seems like a better solution, but would this generate a notification? @InSelfControll needs this to see what the status of the hardware is on their telegram.

With push notifications I get it directly to my teams / telegram as it should be.
I'll keep testing it and update.

**Comment by markdesilva** [comment]
@InSelfControll,

Hi, so from what I understand UK only reports on either status up or down. Once a status is reported, unless it changes (from up to down or down to up) it will not report. UK works that way for all reports. The idea behind this is so UK won't spam you multiple times via email. telegram etc. while you away from your system and can't check to rectify the error.

The only way to keep reporting the critical error is to keep keep flipping the status between up and down. 

When you first report it down, store that some where, when it next reports check the previous status, and if its "down", change your url status to "up" and replace the stored status. The next time it checks the stored status, it will be "up" so then it will  change the status to "down" and so on. For your code:

```
#!/bin/bash

# Get disk usage
disk_usage=`/usr/bin/df -h | /usr/bin/grep "extmedia2" | /usr/bin/awk 'END {print $5}' | /usr/bin/tr -d "%"`

# Check status file and flip status for continuous notices
if [ -f /tmp/du.status ]; then
   if [ `cat /tmp/du.status` == "up" ]; then
        echo "down" > /tmp/du.status
   else
        echo "up" > /tmp/du.status
   fi
else
   echo "up" > /tmp/du.status
fi

udstatus=`cat /tmp/du.status`

# Check if disk usage is higher than 85%
if [ $disk_usage -gt 1 ]; then
  # Send push notification
  /usr/bin/curl -k "https://1.1.1.1:3001/api/push/**********?status=$udstatus&msg=Disk%20Usage%20is%20high:${disk_usage}%25"
fi
```
Take note, this will keep spamming you until you pause the monitor or disable the cron for the script.

Honestly I think the default of only sending the message once is the right way to go. Hope this helps.

@louislam if you have time and you think it is useful, maybe there can be an option in the monitor to allow for "down" status to keep repeating until admin intervenes.

Cheers!

**Comment by InSelfControll** [comment]
Hi,
Thanks for your reply.
I think this option should be in UK that it'll repeat it every X times if the issue didn't fixed.

Let's say it repeats 4 times each time after 1 minute after 3 times the report will changed into critical and the fourth time will be sent via email by the script.
UK will keep reporting every time till it fixed only on issues.
> @InSelfControll,
> 
> Hi, so from what I understand UK only reports on either status up or down. Once a status is reported, unless it changes (from up to down or down to up) it will not report. UK works that way for all reports. The idea behind this is so UK won't spam you multiple times via email. telegram etc. while you away from your system and can't check to rectify the error.
> 
> The only way to keep reporting the critical error is to keep keep flipping the status between up and down. 
> 
> When you first report it down, store that some where, when it next reports check the previous status, and if its "down", change your url status to "up" and replace the stored status. The next time it checks the stored status, it will be "up" so then it will  change the status to "down" and so on. For your code:
> 
> ```
> #!/bin/bash
> 
> # Get disk usage
> disk_usage=`/usr/bin/df -h | /usr/bin/grep "extmedia2" | /usr/bin/awk 'END {print $5}' | /usr/bin/tr -d "%"`
> 
> # Check status file and flip status for continuous notices
> if [ -f /tmp/du.status ]; then
>    if [ `cat /tmp/du.status` == "up" ]; then
>         echo "down" > /tmp/du.status
>    else
>         echo "up" > /tmp/du.status
>    fi
> else
>    echo "up" > /tmp/du.status
> fi
> 
> udstatus=`cat /tmp/du.status`
> 
> # Check if disk usage is higher than 85%
> if [ $disk_usage -gt 1 ]; then
>   # Send push notification
>   /usr/bin/curl -k "https://1.1.1.1:3001/api/push/**********?status=$udstatus&msg=Disk%20Usage%20is%20high:${disk_usage}%25"
> fi
> ```
> Your UK will look like this:
> 
> ![uk-push_flipstatus](https://user-images.githubusercontent.com/59959024/209190840-25005ae6-5572-4e1d-a679-6f6a41fef89c.jpg)
> 
> Take note, this will keep spamming you until you pause the monitor or disable the cron for the script.
> 
> Honestly I think the default of only sending the message once is the right way to go. Hope this helps.
> 
> @louislam if you have time and you think it is useful, maybe there can be an option in the monitor to allow for "down" status to keep repeating until admin intervenes.
> 
> Cheers!
> 
> 
> > With push notifications I get it directly to my teams / telegram as it should be. I'll keep testing it and update.
> > 
> > The issue now that's the monitor get the heart bit but never send the message more than once.
> > 
> > Example: (for the test I did a check that check if the disk_usage is higher than 1% it should send critical alert)
> > 
> > ```shell
> > #!/bin/bash
> > 
> > # Get disk usage
> > disk_usage=`/usr/bin/df -h | /usr/bin/grep "fedora" | /usr/bin/awk 'END {print $5}' | /usr/bin/tr -d "%"`
> > disk_usage1=`/usr/bin/df -h | /usr/bin/grep "fedora" | /usr/bin/awk 'END {print $5}'`
> > 
> > # Check if disk usage is higher than 85%
> > if [ $disk_usage -gt 1 ]; then
> >   # Send push notification
> >   /usr/bin/curl -k "http://1.1.1.1:3001/api/push/********?msg=Disk%20Usage%20is%20high:${disk_usage}%25"
> > fi
> > ```
> > 
> > ![image](https://user-images.githubusercontent.com/20114252/209133979-454800d9-db53-4d28-afd8-a118fc6f9d27.png) Look at the script and the picture.

**Comment by markdesilva** [comment]
> Hi, Thanks for your reply. I think this option should be in UK that it'll repeat it every X times if the issue didn't fixed.
> 
> Let's say it repeats 4 times each time after 1 minute after 3 times the report will changed into critical and the fourth time will be sent via email by the script. UK will keep reporting every time till it fixed only on issues.
> 

Right, so you can do this by setting the resend notifications if Down X times consequently to "4". But like I said, it will not update the UK status, but only keep sending to your alert (telegram, etc).

**Comment by InSelfControll** [comment]
> > Hi, Thanks for your reply. I think this option should be in UK that it'll repeat it every X times if the issue didn't fixed.
> > 
> > Let's say it repeats 4 times each time after 1 minute after 3 times the report will changed into critical and the fourth time will be sent via email by the script. UK will keep reporting every time till it fixed only on issues.
> > 
> 
> Right, so you can _sort of_ do this by setting the resend notifications if Down X times consequently to "4". But like I said, it will not update the UK status, but only keep sending to your alert (telegram, etc).

The issue that if you mark it as down in the url so it'll not send the correct message it just sending "no heartbeat" instead my message.

**Comment by markdesilva** [comment]
Yes, you are right. In the alert (eg: telegram) message, it will only say "No heartbeat in the time window". For your own message to appear in the alert message, you will need to use my modifications to your script, just that the status will keep flipping between up and down.

> The issue that if you mark it as down in the url so it'll not send the correct message it just sending "no heartbeat" instead my message.

**Comment by InSelfControll** [comment]
> Yes, you are right. In the alert (eg: telegram) message, it will only say "No heartbeat in the time window". For your own message to appear in the alert message, you will need to use my modifications to your script, just that the status will keep flipping between up and down.
> 
> > The issue that if you mark it as down in the url so it'll not send the correct message it just sending "no heartbeat" instead my message.

Hey look I have another script that works really nice and with custom status=down message but the only issue now that for some reason it doesn't get %25 (%) symbol at the end.

```bash
!/bin/bash

# This script monitors RAM, CPU, and disk usage and sends an alert if disk usage is higher than 85%.

# Get current disk usage
DISKUSAGE=`/usr/bin/df -h | /usr/bin/awk '$NF=="/"{printf "%s\t\t", $5}' | /usr/bin/tr -d "%"`
# Get current RAM and CPU usage
RAM=`free -m | awk 'NR==2{printf "Memory Usage: %s/%sMB (%.2f%%)\n", $3,$2,$3*100/$2 }'`
CPU=`top -bn1 | grep load | awk '{printf "CPU Load: %.2f\n", $(NF-2)}'`
# Check if disk usage is higher than 85%
if [[ ${DISKUSAGE%?} -gt 21 ]]; then
  echo "High Disk Usage: $DISKUSAGE"
  echo "$RAM"
  echo "$CPU"
  # Send alert
  curl -s "https://uk.***com/api/push/******?status=down&msg=Disk%20usage%20is%20high:${DISKUSAGE}%25"
  else
          curl -s "https://uk.****.com/api/push/******?status=up&msg=Disk%20usage%20is%20Fixed:${DISKUSAGE}%25"
fi
```
I would be very happy if we can fix it together :)

**Comment by markdesilva** [comment]
The problem is that there is a trailing newline (/n) on your DISKUSAGE variable.

if you put [ ] in front and behind the DISKUSAGE variable when you echo it (line 12), you will see.

`echo "High Disk Usage: [$DISKUSAGE]"`

If you echo the curl url in your code you will also see:

`echo "https://uk.***com/api/push/******?status=down&msg=Disk%20usage%20is%20high:${DISKUSAGE}%25"`

Also the first line should be `#!/bin/bash` not `!/bin/bash`

Cheers!


> Hey look I have another script that works really nice and with custom status=down message but the only issue now that for some reason it doesn't get %25 (%) symbol at the end.
> 
> ```shell
> !/bin/bash
> 
> # This script monitors RAM, CPU, and disk usage and sends an alert if disk usage is higher than 85%.
> 
> # Get current disk usage
> DISKUSAGE=`/usr/bin/df -h | /usr/bin/awk '$NF=="/"{printf "%s\t\t", $5}' | /usr/bin/tr -d "%"`
> # Get current RAM and CPU usage
> RAM=`free -m | awk 'NR==2{printf "Memory Usage: %s/%sMB (%.2f%%)\n", $3,$2,$3*100/$2 }'`
> CPU=`top -bn1 | grep load | awk '{printf "CPU Load: %.2f\n", $(NF-2)}'`
> # Check if disk usage is higher than 85%
> if [[ ${DISKUSAGE%?} -gt 21 ]]; then
>   echo "High Disk Usage: $DISKUSAGE"
>   echo "$RAM"
>   echo "$CPU"
>   # Send alert
>   curl -s "https://uk.***com/api/push/******?status=down&msg=Disk%20usage%20is%20high:${DISKUSAGE}%25"
>   else
>           curl -s "https://uk.****.com/api/push/******?status=up&msg=Disk%20usage%20is%20Fixed:${DISKUSAGE}%25"
> fi
> ```
> 
> I would be very happy if we can fix it together :)

**Comment by InSelfControll** [comment]
> @InSelfControll 
> 
> The problem is that there is a trailing newline (/n) on your DISKUSAGE variable.
> 
> if you put [ ] in front and behind the DISKUSAGE variable when you echo it (line 12), you will see.
> 
> `echo "High Disk Usage: [$DISKUSAGE]"`
> 
> If you echo the curl url in your code you will also see:
> 
> `echo "https://uk.***com/api/push/******?status=down&msg=Disk%20usage%20is%20high:${DISKUSAGE}%25"`
> 
> Also the first line should be `#!/bin/bash` not `!/bin/bash`
> 
> Cheers!
> 
> 
> > Hey look I have another script that works really nice and with custom status=down message but the only issue now that for some reason it doesn't get %25 (%) symbol at the end.
> > 
> > ```shell
> > !/bin/bash
> > 
> > # This script monitors RAM, CPU, and disk usage and sends an alert if disk usage is higher than 85%.
> > 
> > # Get current disk usage
> > DISKUSAGE=`/usr/bin/df -h | /usr/bin/awk '$NF=="/"{printf "%s\t\t", $5}' | /usr/bin/tr -d "%"`
> > # Get current RAM and CPU usage
> > RAM=`free -m | awk 'NR==2{printf "Memory Usage: %s/%sMB (%.2f%%)\n", $3,$2,$3*100/$2 }'`
> > CPU=`top -bn1 | grep load | awk '{printf "CPU Load: %.2f\n", $(NF-2)}'`
> > # Check if disk usage is higher than 85%
> > if [[ ${DISKUSAGE%?} -gt 21 ]]; then
> >   echo "High Disk Usage: $DISKUSAGE"
> >   echo "$RAM"
> >   echo "$CPU"
> >   # Send alert
> >   curl -s "https://uk.***com/api/push/******?status=down&msg=Disk%20usage%20is%20high:${DISKUSAGE}%25"
> >   else
> >           curl -s "https://uk.****.com/api/push/******?status=up&msg=Disk%20usage%20is%20Fixed:${DISKUSAGE}%25"
> > fi
> > ```
> > 
> > I would be very happy if we can fix it together :)
> 
> 

It didn't copied the first line 😆 my script have it already with #!/bin/bash,

**Comment by markdesilva** [comment]
> It didn't copied the first line 😆 my script have it already with #!/bin/bash,

Oh ok. Also its trailing tabs not newline, my mistake.

**Comment by InSelfControll** [comment]
> > It didn't copied the first line 😆 my script have it already with #!/bin/bash,
> 
> Oh ok. Also its trailing tabs not newline, my mistake. 
> 

Yeah printf adding tabs, I removed it but still curl doesn't get %25 ASCII

**Comment by markdesilva** [comment]
> Yeah printf adding tabs, I removed it but still curl doesn't get %25 ASCII

Funny, it works for me.

This is my modified version of your script:

```
#!/bin/bash

# This script monitors RAM, CPU, and disk usage and sends an alert if disk usage is higher than 85%.

# Get current disk usage
DISKUSAGE=`/usr/bin/df -h | /usr/bin/awk '$NF=="/" {printf "%s", $5}' | /usr/bin/tr -d "%"`
# Get current RAM and CPU usage
RAM=`free -m | awk 'NR==2{printf "Memory Usage: %s/%sMB (%.2f%%)\n", $3,$2,$3*100/$2 }'`
CPU=`top -bn1 | grep load | awk '{printf "CPU Load: %.2f\n", $(NF-2)}'`
# Check if disk usage is higher than 85%

if [ ${DISKUSAGE} -gt 1 ]; then
  echo "High Disk Usage: [$DISKUSAGE]"
  echo "$RAM"
  echo "$CPU"
  # Send alert
  curl -s "https://uk.***com/api/push/******?status=up&msg=Disk%20usage%20is%20high:${DISKUSAGE}%25"
else
  curl -s "https://uk.***com/api/push/******??status=up&msg=Disk%20usage%20is%20Fixed:${DISKUSAGE}%25"
fi
```
Did you forget your 3001 or are you running your UK server on 443?

**Comment by InSelfControll** [comment]
> @InSelfControll 
> 
> > Yeah printf adding tabs, I removed it but still curl doesn't get %25 ASCII
> 
> Funny, it works for me.
> 
> This is my modified version of your script:
> 
> ```
> #!/bin/bash
> 
> # This script monitors RAM, CPU, and disk usage and sends an alert if disk usage is higher than 85%.
> 
> # Get current disk usage
> DISKUSAGE=`/usr/bin/df -h | /usr/bin/awk '$NF=="/" {printf "%s", $5}' | /usr/bin/tr -d "%"`
> # Get current RAM and CPU usage
> RAM=`free -m | awk 'NR==2{printf "Memory Usage: %s/%sMB (%.2f%%)\n", $3,$2,$3*100/$2 }'`
> CPU=`top -bn1 | grep load | awk '{printf "CPU Load: %.2f\n", $(NF-2)}'`
> # Check if disk usage is higher than 85%
> 
> if [ ${DISKUSAGE} -gt 1 ]; then
>   echo "High Disk Usage: [$DISKUSAGE]"
>   echo "$RAM"
>   echo "$CPU"
>   # Send alert
>   curl -s "https://uk.***com/api/push/******?status=up&msg=Disk%20usage%20is%20high:${DISKUSAGE}%25"
> else
>   curl -s "https://uk.***com/api/push/******??status=up&msg=Disk%20usage%20is%20Fixed:${DISKUSAGE}%25"
> fi
> ```
> Did you forget your 3001 or are you running your UK server on 443?
> 
> 

This is my private server the other one is stg at work with port 3001.
This specific url run behind treafik proxy on docker :) so no port.

**Comment by markdesilva** [comment]
Ah ok. In anycase the %25 works for me.

For CPU % used, have you tried using iostat (apt get sysstat) and piping that into bc?

```
CPUP_IDLE=`iostat  | grep -A1 "avg-cpu" | awk {'print $6'} | tail -1`
CPUP_USED=`echo "100 - $CPUP_IDLE" | bc`
```

For memory, `free -g` take the available/total and also pipe into `bc`.

> This is my private server the other one is stg at work with port 3001. This specific url run behind treafik proxy on docker :) so no port.

**Comment by InSelfControll** [comment]
> Ah ok. In anycase the %25 works for me.
> 
> For CPU % used, have you tried using iostat (apt get sysstat) and piping that into bc?
> 
> ```
> CPUP_IDLE=`iostat  | grep -A1 "avg-cpu" | awk {'print $6'} | tail -1`
> CPUP_USED=`echo "100 - $CPUP_IDLE" | bc`
> ```
> 
> For memory, `free -g` take the available/total and also pipe into `bc`.
> 
> > This is my private server the other one is stg at work with port 3001. This specific url run behind treafik proxy on docker :) so no port.
> 
> 

Done!
Thanks for your help.
I think that UK need to get this feature that send critical messages on hosts if there are issues other than the server is down.
Like that the http test that can monitor string and still get code:200 but mark as missing string.

I would love to see UK get this feature in the next release 😎

**Comment by InSelfControll** [comment]
I just found an issue while running it via docker and on a VM with nodejs.
It keep losing connection every time and then sending service down - "no heartbeat" because the connection is "down" even when the service is up and running correctly.

**Comment by BasToTheMax** [comment]
Any updates? Would love to have this.

**Comment by milzamsz** [comment]
found existing solution https://github.com/msgbyte/tianji

**Comment by MichaelBelgium** [comment]
@milzamsz If i'm correct, on tianji you're just having a list of servers. I don't see the possiblity to, for example, set alerts when cpu usage > x % 

Neither add a server to a status page? And doesn't support mysql?

**Comment by milzamsz** [comment]
@MichaelBelgium yes but it's better for me who doesn't need advanced monitoring like netdata/prometheus. im installing with easypanel and it come with postgres

**Comment by AquaMCU** [comment]
Hi. Id like to join the conversation ;)

Other idea: How about KUMA just adds a new monitor, that requests a page from the server to be monitored. Here you guys can knock yourself out and implement your GO_JS_PHP_BASH_FORTRAN or whatever module, that just responds with a percentage and does not respond when it s critical.

As for KUMA, for this Monitor, just display the data and make it NOK when it is not responding.

... easy to do for the KUMA team and nice and hackable for the rest of us ;)

Oliver

**Comment by andrijs29** [comment]
is this will make uptime kuma like hetrix tools?
for me its good if UK can have something like cpu/network/ram/disk/ usage like hetrix tools

im currently using hetrix and its help to show the current VPS usage without login to my SSH 
what happened before im login to my terminal

**Comment by apio-sys** [comment]
This is about Uptime as the name of the project. Not resource monitoring. Those are different issues and have their own tools. You might use Uptime to show to a client that SLA is respected and you might have internal perf monitoring tools so you can adjust power as needed to customer sees their SLA respected. That's just MHO and I really think you shouldn't mix this up or you'll end up with a tool that really doesn't fit any need anymore.

**Comment by GAS85** [comment]
Didn't read whole tread, but it is pretty easy to configure: https://github.com/GAS85/Uptime-Kuma-ram-disk-cpu

**Comment by bobbles911** [comment]
literally the only thing stopping me using UK has been this.

I don't need fancy graphs, all I need to know is if the disk is > 90% (or the RAM or CPU).
Really curious to know what kind of people are concerned with uptime yet *don't* need this feature... Have you really never shipped code with a memory leak, and do you really want to find out only *After* the server is no longer available and can't even be ssh'd into because everything has broken?

(And yes... I do want this feature so I can stop paying a lot for a commercial system, when UK already implements 95% of what they offer for free!)

I will try something along the lines of @GAS85 suggestion but it would be so nice if it was just a script you can copy + paste from the UK console and it self installs and "just works".

**Comment by markdesilva** [comment]
> I don't need fancy graphs, all I need to know is if the disk is > 90% (or the RAM or CPU). Really curious to know what kind of people are concerned with uptime yet _don't_ need this feature... Have you really never shipped code with a memory leak, and do you really want to find out only _After_ the server is no longer available and can't even be ssh'd into because everything has broken?

The kind of people who understand what the purpose of the FREE software is when they decided to use it and read about its features and learn how to use those features. PUSH has been a feature since day one. 

> I will try something along the lines of [@GAS85](https://github.com/GAS85) suggestion

This is just a PUSH, which again, has been a feature since day 1. Mind you, you will need to edit the script and fill in the token and UK URL. You know, the one that's generated for you when you create the push monitor type? And you do know that the scripts @GAS85 provided has to run on your server right?

> but it would be so nice if it was just a script you can copy + paste from the UK console and it self installs and "just works".

Well you are a developer, why don't you write one that does what you want? Or at the very least fork UK and add to it whatever you need for yourself.

This is a software, provided to us for free and some of us are behaving like the developer of said software is indebted to us somehow to have it do all their whims and fancies when it is not even the purpose of the software in the first place.

**Comment by CommanderStorm** [comment]
> Quick reminder for everybody: Issues are for discussing what needs to be done how by whom. [We use 👍🏻 on issues to prioritise some work](https://github.com/louislam/uptime-kuma/issues?q=is%3Aissue+is%3Aopen+sort%3Areactions-%2B1-desc), as always: [Pull Requests](https://github.com/louislam/uptime-kuma/blob/master/CONTRIBUTING.md) welcome.
> 
> You can currently implement such a monitor without any modification needed via the post monitor as stated above. If you want to simplify this process, adding a monitor via a PR is a possibility (see our contribution guide for additional details).

Let's keep the discussion respectful and constructive. 🐻

Improved support for this use case would certainly be nice to have. At the moment, however, I don’t have the capacity to implement it myself - there are many feature requests and only limited maintainer time from my end (currently writing my masters thesis).

In the meantime, this can be achieved via the system services monitor, push monitor, or other existing approaches. If you want a beter integrated solution, contributions in that direction would be greatly appreciated.
I can review some amount of PRs, even though at least for the next month not at the rate I might want.

**Comment by markdesilva** [comment]
> > Quick reminder for everybody: Issues are for discussing what needs to be done how by whom. [We use 👍🏻 on issues to prioritise some work](https://github.com/louislam/uptime-kuma/issues?q=is%3Aissue+is%3Aopen+sort%3Areactions-%2B1-desc), as always: [Pull Requests](https://github.com/louislam/uptime-kuma/blob/master/CONTRIBUTING.md) welcome.
> > You can currently implement such a monitor without any modification needed via the post monitor as stated above. If you want to simplify this process, adding a monitor via a PR is a possibility (see our contribution guide for additional details).
> 
> Let's keep the discussion respectful and constructive. 🐻
> 
> Improved support for this use case would certainly be nice to have. At the moment, however, I don’t have the capacity to implement it myself - there are many feature requests and only limited maintainer time from my end (currently writing my masters thesis).
> 
> In the meantime, this can be achieved via the system services monitor, push monitor, or other existing approaches. If you want a beter integrated solution, contributions in that direction would be greatly appreciated. I can review some amount of PRs, even though at least for the next month not at the rate I might want.

Apologies, didn't realize this is "abuse", given the nature of the post I was replying to. I still don't see how it is "abuse", but I will refrain from replying.

Cheers.

**Comment by CommanderStorm** [comment]
Not really abuse. That is a bit of a harsh word.
It is just the most fitting from the amount of hide-options (spam, abuse, off-topic, outdated, duplicate, resolved) that I can choose of ^^

I would like this issue to be in a more newcomer/contribution friendly and constructive tone, that is all.
No bad feelings.

**Comment by GAS85** [comment]
Out of curiosity, how is Uptime Kuma expected to implement this request on its own? Data collection is not always trivial, and there’s no universal solution—especially when it comes to remote hosts. For example, looking at [Cacti](https://github.com/Cacti/cacti), it can collect data from remote servers, but it does so via protocols like SNMP. Under the hood, there is still a script or polling mechanism that gathers the data and is periodically executed by Cacti.

So should Kuma implement script-based monitoring? That approach wouldn’t work very well in containerized environments or when dealing with remote hosts.

I think a push monitor combined with scripts running on the local or remote host is a more flexible and efficient solution. You can decide locally when to trigger an alert (send a “down” message) and when not to, and you can tailor the monitoring logic to your specific environment.

**Comment by markdesilva** [comment]
> Out of curiosity, how is Uptime Kuma expected to implement this request on its own? Data collection is not always trivial, and there’s no universal solution—especially when it comes to remote hosts. For example, looking at [Cacti](https://github.com/Cacti/cacti), it can collect data from remote servers, but it does so via protocols like SNMP. Under the hood, there is still a script or polling mechanism that gathers the data and is periodically executed by Cacti.
> 
> So should Kuma implement script-based monitoring? That approach wouldn’t work very well in containerized environments or when dealing with remote hosts.
> 
> I think a push monitor combined with scripts running on the local or remote host is a more flexible and efficient solution. You can decide locally when to trigger an alert (send a “down” message) and when not to, and you can tailor the monitoring logic to your specific environment.

Yes push is almost universal in that aspect, regardless of OS or protocol. If UK allows us to define our own units for the X and Y axis of the graph as I mentioned some time ago, it would be ideal.

**Comment by GAS85** [comment]
Ahhh, this could be cool, because I already use response time to track e.g. my backups speed (it could be delivered with Push API Call as value):

<img width="969" height="250" alt="Image" src="https://github.com/user-attachments/assets/57d9a895-3213-4ea3-a92a-d086258b7d93" />

For relative (percentage) it fits quite well, e.g.: RAM usage:

<img width="1088" height="250" alt="Image" src="https://github.com/user-attachments/assets/945ae8b9-1cd1-4d44-aefd-dfb8f6ade5a6" />

I simply "know" that for RAM, CPU, etc, it is `%` and for everything else it is in `ms`

**Comment by mustafaulukaya** [comment]
You can use this agent to monitor specific system vitals via HTTP. Because Uptime Kuma supports HTTP-based health checks and allows response handling, you can set up alerts such as triggering a notification when RAM usage goes above 80%.


https://github.com/appouse/HostPeek/

**Comment by krupper** [comment]
I think the core principle of UK is the simplicity of monitoring services, whereas many other projects require building extensive dashboards and have a steep learning curve. 

With that in mind, I believe it would make sense to include just a few simple metrics that provide a quick overview. They would help quickly rule out certain factors in the event of an outage. However, I feel these metrics do not belong on a public status page but are intended for internal use only. 

Specifically, I would suggest integrating the following metrics (all measured in percentages):
- Free disk space
- iNodes
- RAM utilization
- CPU utilization
- Network utilization

For all metrics, UK should only provide API endpoints and the UI. The data should be sent to the UK instance via an agent. By streamlining and focusing on a good UI, the core principle of UK should still be preserved.
